batch/batch/cloud/gcp/worker/worker_api.py

```diff
@@ -128,3 +138,111 @@ async def close(self):
 
     def __str__(self):
         return f'project={self.project} zone={self.zone}'
+
+
+ContainerCredentials = Dict[str, aiogoogle.GoogleServiceAccountCredentials]
+
+
+class GoogleHailMetadataServer(HailMetadataServer[GCPUserCredentials, ContainerCredentials]):
+    def __init__(self, project: str, http_session: httpx.ClientSession):
+        super().__init__()
+        self._project = project
+        self._metadata_server_client = aiogoogle.GoogleMetadataServerClient(http_session)
+
+    def _create_container_credentials(self, default_credentials: GCPUserCredentials):
+        default_sa_creds = aiogoogle.GoogleServiceAccountCredentials(orjson.loads(default_credentials.key))
+        credentials = {}
+        for name in ('default', default_sa_creds.email):
+            credentials[name] = default_sa_creds
+        return credentials
+
+    async def _close_container_credentials(self, container_credentials: ContainerCredentials):
+        for credential in container_credentials.values():
+            await credential.close()
+
+    def _container_credentials(self, request: web.Request) -> ContainerCredentials:
+        return request['credentials']
+
+    def _user_credentials(self, request: web.Request) -> aiogoogle.GoogleServiceAccountCredentials:
+        email = request.match_info.get('gsa') or 'default'
+        return self._container_credentials(request)[email]
+
+    async def root(self, _):
+        return web.Response(text='computeMetadata/\n')
+
+    async def project_id(self, _):
+        return web.Response(text=self._project)
+
+    async def numeric_project_id(self, _):
+        return web.Response(text=await self._metadata_server_client.numeric_project_id())
+
+    async def service_accounts(self, request: web.Request):
+        accounts = '\n'.join(self._container_credentials(request).keys())
```

Comment:
Should the accounts have trailing slashes like the root does? Should the root not have a trailing slash?

Reply:
Not sure exactly what you mean by this comment, but the pattern I've seen is that any directory listings (ending in a trailing slash) list nested endpoints without trailing slashes.